### ISSUE TITLE
Feature request: Option for auto-approve and auto-bonus

### ISSUE THREAD

**Issue by fredcallaway** [NYUCCL/psiTurk#286]
We always approve and assign pre-computed bonuses to everyone who completes our experiments. Thus, I would like to avoid opening up psiturk to "manually" approve and bonus. Would it be possible to have a config option for automatically approving and bonusing on completion?

**Comment by deargle** [comment]
Probably possible, but not necessarily easy.

Here's some convenience bash scripts I use for approving and bonusing (bonusing one needs work). Requires installing `jq`

``` bash
# This script reads a list of submitted_workers, and approves each one.
set -x
psiturk -e "worker list --submitted" | jq -c '.[] | .assignmentId' | while read i; do # `.[]` returns each object in the array, and .assignmentId extracts the assignmentId for each
    psiturk -e "worker approve ${i//\"}" # remove quote marks from assignment_id
done
```

``` bash
set -x
psiturk -e "worker list --approved" | jq -c '.[] | .hitId' | uniq | while read i; 
    psiturk -e "worker bonus --auto --hit ${i//\"}"
done
```

**Comment by deargle** [comment]
@fredcallaway I just updated my scripts to be slicker. 
* `-r` in `jq` strips quotes for you
* build a list of hitIds of workers just approved, and then feed that list into the bonus script.

**Comment by fredcallaway** [comment]
The conciseness of that first script makes me want to learn bash scripting.

That being said, it's kind of a hack. I think the "right" thing would be a python script that interfaces directly with the psiturk library. At present, there doesn't appear to be a public python API, but `PsiturkNetworkShell` could be adapted to serve this purpose.

**Comment by deargle** [comment]
Indeed it is hacky. But I like it better than the hacking I was doing before, [see this gist from one of my projects](https://gist.github.com/deargle/b15272fe2a883ec7bbf68f9b131edc64). I call amt_services functions directly, and manually set db status.

**Comment by fredcallaway** [comment]
How do we feel about refactoring `PsiturkNetworkShell` to separate the backend logic from the shell interface (e.g. printing)? A cursory read through the code makes me think this would be fairly easy. I can take a stab at it.

**Comment by deargle** [comment]
Stab away!

**Comment by gureckis** [comment]
@fredcallaway been a dream for a while so we can script things instead of always relying on the shell.  time and energy for sorting out a refactor has been the obstacle.  the proposal i always had in mind was to rewrite the shell exactly as you say and then provide a psiturk library (i.e., `import psiturk`) for scripting directly within python.

**Comment by fredcallaway** [comment]
Reviving this issue now that we're in post-conference-deadline season. I think a reasonable (albeit ambitious) way forward is to address this issue along with the Python 3 issue with a new major version (PsiTurk 3).

I could take the lead on this, but I would want to drop Python 2 support. Additionally, I wouldn't want to re-implement the shell because I can't think of a case that I would rather use the shell than a script, jupyter notebook, or IPython terminal. 

I might initially aim to support external server deployment (e.g. heroku) only (i.e. skip gunicorn) but I think it would be good to bring that back before a public release. Perhaps it makes more sense for me to split this off into a separate project. But I wanted to check to see what people here think.

**Comment by deargle** [comment]
Fyi moving to python 3 involves moving to boto3, which uses a different
(more current) mturk api -- so the only code replication python-wise would
be the server, not the shell functionality or mturk api calls.

**Comment by jacob-lee** [comment]
Moving to python3 is probably inevitable. boto might not be supported much longer, the cmd2 library is dropping support for python27, and so on. Some of this will require substantial work; might as well get ahead of it. boto3 will probably be a pain point because the API is completely different.

Major changes to psiturk must answer the question, "what value does psiturk provide over what one can already do with boto3, etc."

I like the idea of a importable psiturk library for python scripting, but I think it would be a mistake to eliminate other methods of interacting with psiturk. The interactive shell should stay: I think people use the interactive shell, even if power users tend not too. However, it should no longer be the primary interface to its functionality: there is not reason that psiturk commands should be interpreted by cmd2. A command line interface should leverage python's argparse, unless called without arguments, in which case an interactive interpreter is loaded.

The network shell interface needs a major refactor. psiturk can be run in some many ways (interactive, script, statement execute), and in so many modes: (cabin mode, debug mode, sandbox mode, live), but most of these are handled by the same network shell object (so lots of branching in the code). Its hard to maintain, and bound to get harder the more features that are added, unless its separated somehow. One obstacle to doing that, however, is the decision to support changing config changes/mode changes in the same session. The current interactive shell would have to return control to a shell manager, somehow.

Removing support for self-hosted servers would be a big mistake.

**Comment by fredcallaway** [comment]
> Major changes to psiturk must answer the question, "what value does psiturk provide over what one can already do with boto3, etc."

This is a good way to think. For me, the main value of psiturk is that it handles the server/client interaction (e.g. condition assignment, repeat participant blocking, and saving data). The other primary features of psiturk (that I can identiy) are server hosting (with gunicorn) and AWS interaction (a wrapper around boto).

I agree that the server hosting is an important feature that should not be dropped; however I don't use this aspect of psiturk so I'm not motivated to work on porting it to Python 3.

I think the AWS interaction is the primary weakness of the current version of psiturk. Creating HITs and compensating workers involves to much manual effort, and I would like it to be scriptable. This is something that I'm willing to put a fair amount of effort into.

**Comment by deargle** [comment]
Here's my value-add:
- server/client interaction (e.g. condition assignment, repeat participant blocking, and saving data)
- flask/gunicorn stack* 
-- for debugging locally (@fredcallaway, you say you don't use the gunicorn functionality at all -- you don't debug locally?)
-- for use in heroku deploys, etc, which still require something like gunicorn in front of Flask to serve the routes.
- Boto wrapper is good for convenience and for grad students afraid of scripting

Problem children:
- the psiturk ad server. Users are over-dependent on it because they don't know how to do SSL cert management. But if something like Heroku is used, SSL is built in.

I don't see a strong need to keep the shell around, even for non-power users. In theory, all shell functionality and syntax can be replicated outside of the shell. E.g., `hit create` becomes `psiturk hit create`, `worker list` becomes `psiturk worker list`. What's lost by dropping the shell, besides the prompt showing which `mode` you're in and `server status` and saving a bit of typing? Code-completion and interactive prompts can be implemented outside of the shell.

I'm wholly behind enabling direct scripting via `import psiturk`. The **only** reason that the psiturk wrapper around boto is essential is because it enables psiturk to keep its own db status in sync for workers.

For that matter, will it really be that much work to abstract the current shell commands one layer and have shell commands call psiturk library commands? With that done, direct scripting would be done. For my puny mind, that's a quick win before full conversion to python 3, and doing it now could help prepare for the migration.

\* but it's still crucial that nginx be put in front to serve static files. Gunicorn slows to a crawl on its own for all my stuff on debug, but heroku's default buffering seems to be sufficient even without an nginx added into the mix (but I plan to add nginx soon to my stuff)

**Comment by gureckis** [comment]
I agree with @deargle that if you abstract the current shell commands into one python API (ala `import psiturk`) you'll accomplish two things at once.  one is the library would be started.  second is that you expose a bunch of easier ways to use psiturk.  this can be the interactive shell, a command line declarative type interface, and even a web-based interface (because you can add a /admin route to the server and enable a gui).  The GUI approach is probably better than the command line because it can be used for things hosted on remote servers more easily than using `screen` etc...  We had a dashboard and scrapped it because we weren't sure all the commands we would want to add.  We figured adding commands is easier to a command line interface and then when things stabilized it could be abstracted into something *really* user friendly.  anyway, i'm happy there's some interest in pushing on this and happy to help as much as i can!

**Comment by jacob-lee** [comment]
I concur. A general API to psiturk services would make the design of client interfaces (gui, cli, web interface, etc) much cleaner, by separating how a service is called from what the service provides (case in point: the repeated checks of the `quiet` flag in PsiturkNetworkShell.

Many of Dave's examples of added value (condition assignment, repeat participant blocking, and saving data), are things that have made designing our own experiments harder than they might otherwise have been. Nonetheless, I recognize that psiturk provides an accessible entry point into running mturk experiments because it makes good, solid choices for the user, so that they don't have to reinvent the wheel, and get started with their real business. 

I just wish more of those choices were only defaults, defaults that could be over-ridden without forking psiturk (as I have had to). To do that, psiturk would need to become much more loosely coupled than it is now, but not so loosely coupled that it becomes impossible to continue to the provide a uniform client interface to these services. That's a trick.

**Comment by deargle** [comment]
FWIW lest anyone else start I've nearly completed an abstraction of amt_services from the shell -- only took an hour or so. Will share soon on a branch.

**Comment by fredcallaway** [comment]
If we went the GUI web interface route, would you need to have config variables stored in a server-side database? Or could you read and write to a plain text config file (maybe use json or yaml instead of the `.ini` syntax of config.txt)?

**Comment by deargle** [comment]
Okay here it is. https://github.com/NYUCCL/psiTurk/tree/shell-mturk-abstraction

New file called `amt_services_wrapper` that can be imported by calling `import psiturk.amt_services_wrapper` in any python script.

Tested pretty thoroughly all shell commands to make sure they still work. Haven't tested by direct import into a python script yet though. Going to bed 😴 

And @fredcallaway anything is possible for configs. A GUI would just call url api endpoint routes as specified in something like custom.py, which loads configs just like psiturk file. Path of least resistance is to keep it as-is, would work fine for a gui.

**Comment by jacob-lee** [comment]
Great. I think the next step is to remove what is properly the concern of the client, e.g. removing print statements, or checking whether its in quiet mode, etc. For example, instead of:

```python
    def worker_reject(self, chosen_hit, assignment_ids = None):
        ''' Reject worker '''
        if chosen_hit:
            workers = self.amt_services.get_workers("Submitted")
            assignment_ids = [worker['assignmentId'] for worker in workers if \
                              worker['hitId'] == chosen_hit]
            print 'rejecting workers for HIT', chosen_hit
        for assignment_id in assignment_ids:
            success = self.amt_services.reject_worker(assignment_id)
            if success:
                print 'rejected', assignment_id
            else:
                print '*** failed to reject', assignment_id
```

we do (something like) this:

```python
    def worker_reject(self, hit_ids=[], assignment_ids=[]):
        ''' Reject worker '''
        response = {}
        for hit_id in hit_ids:
            workers = self.amt_services.get_workers('Submitted')
            assignment_ids = assignment_ids + [worker['assignmentId'] for worker in workers if \
                              worker['hitId'] == chosen_hit]
        for assignment_id in assignment_ids:
            response[assignment_id] = self.amt_services.reject_worker(assignment_id)
        return response
```

**Comment by deargle** [comment]
@jacob-lee True that we'll need to not use `print` statements if we're using a web GUI. Meanwhile though this should work for scripting purposes.

And `quiet` is currently only used in one place -- in deciding whether to show a confirmation dialog for `hit create`. I'm not sure how to abstract that into the shell code, unless we just remove it entirely. We will also need to move any `raw_input` calls to the shell -- but those are skipped if all parameters are provided for `hit create`, so again this is good to go for scripting.

> An aside: this way of finding submitted workers for particular hit id seems really slow

You're right. I've already optimized a few other places in just the way you describe for selfish reasons (for my own scripts), but there are other places like where you cited that need to be fixed still. I've yet to reject anyone.

**Comment by jacob-lee** [comment]
Could be that we are thinking along different tracks. I don't see a place
for any kind confirmation dialogue, or other interactive content in a
services api; nor print statements. If a command doesn't have all the
information it needs to execute, it should throw an exception to be handled
by the client (or return an object indicating the failure). That client can
be a web gui, a CLI, commands in a bash script, or a python script that
directly imports the api. All of those are great, but pose different
requirements that are best handled independently.

**Comment by deargle** [comment]
I got you. What I meant though is that a minimum viable product was ready to be used for scripting even though the amt_wrapper still has print statements in it. I worked some more on this tonight and made some progress (already pushed to the branch):

- worker_list in `amt_services_wrapper` now returns json instead of immediately printing json. The shell immediately prints the json.
- the cli is now faster -- instead of having to call `psiturk -e "amt_balance"`, you can now call commands a la `psiturk amt_balance` or `psiturk hit create 1 0.01 1`. All unknown args are assumed to be shell commands.
- I'm using `quiet` all over the place in the shell again, but no longer at any place in the `amt_wrapper`. With the use of `quiet` in the shell, load times are much better. Also, I'm now lazy-loading services like `psiturk_org_services` so that they're not loaded immediately at shell init, and in the case of psiturk_org_services, not at all if the ad server isn't being used. So timeout errors won't be an issue there anymore. It's not needed for approving workers. The shell still makes a call to `psiturk_org_services` for "news" if a traditional interactive shell is launched.
- `amt_services_wrapper` lazy-loads all services, one outcome of which is that we won't get the annoying message about being unable to connect to aws rd (database services) which I don't think anyone uses anyway, which is funny since over half of the amt_services_wrapper code is dedicated to aws rd calls.
- interactivity for `hit_create` was moved out of `amt_services_wrapper` and into the shell code. I added some crude general `raise Exception` statements in there too. A lot of, but not all, of printing for that function was moved into the shell as well (and therefore into the cli).

That's a side note -- any cli commands instantiate a shell object, albeit with `quiet` passed. One day we can create another shell object specific to scripting, but not this day. `Quiet` is tacky, but good enough for now. Although it's not unheard of to pass quieting arguments to scripts...

Here's what I meant by a minimum viable product for scripting being ready -- `worker list` already returns json. That can be looped over to make calls to `worker approve` and `worker bonus`, which are the two big things that started this thread. This can now be done in python by `import psiturk.amt_services_wrapper`, and it can still be done in bash, more elegantly now without having to do `-e` (but the functionality is the same). `hit list` still needs to return json, and yes, all print statements need to be removed, but it works as-is.

**Comment by jacob-lee** [comment]
I was also working on my own version of the wrapper last night. However, there is no point in duplicating effort. Let me know how I can help.

I have two suggestions:

1. that we move towards having these methods return data encapsulated in dicts, because a relatively thin wrapper can readily convert that to json for data interchange with web clients, but is also readily used by other kinds of client code, and because it lets us add metadata that clients might find useful.

2. that we give clients a uniform way to detect when something has gone wrong. Either with a class (hierarchy) of PsiturkAPIExceptions to be caught, or by an error field in the returned dict.

Also, eventually adding logging for command execution might be a good idea.

**Comment by deargle** [comment]
Let's figure out how we can work together on this. Are you comfortable working off of my branch? It's by no means finished, I'm just incrementally improving it. I think the general structure is good though -- i.e., all `amt_service` calls have been moved out of the shell and into that class. 

What do you want to work on, the exception hierarcy and returning dicts? Whatever we do, I think for now that the goal should be to have the shell and cli calls behave in exactly the way they currently do, so the abstraction is transparent to users. 

Here's the pattern I'm doing:

shell/cli does argparse via cmd2:

https://github.com/NYUCCL/psiTurk/blob/7ae468c809ff7f0982e937e046b657930d5c560c/psiturk/psiturk_shell.py#L918-L931

Which calls its own version of the command with an identical signature (room for some kind of interface here to help maintain signatures, got any ideas there?). This then eventually calls the wrapper version of the function, after doing interactive input checking. That wrapper call is wrapped in a try-except block which currently does just bare-bones general Exception handling.

https://github.com/NYUCCL/psiTurk/blob/7ae468c809ff7f0982e937e046b657930d5c560c/psiturk/psiturk_shell.py#L222-L283

If no exception was thrown, The wrapper return some data (which we can make into a dict), and then the shell continues with printing .

One thing to figure out is when to make the shift to Python 3. I'm not familiar with how much things like exception handling will change -- if we want to upgrade first and then implement better exception handling in the wrapper, or if it's fine to go ahead and implement better exception handling in 2.7 and then upgrade later. Same for logging -- unsure how much that will change. I don't expect much. So probably the answer is to finish the wrapper/api in 2.7, and then shift. At least all Boto calls are already in their own class in psiturk.

> 1. that we move towards having these methods return data encapsulated in dicts

Agreed. I misspoke when I said "hit list still needs to return json", meant "something parse-able" instead of just the printed list that it currently does. Won't be a hard fix.

> 2.  that we give clients a uniform way to detect when something has gone wrong.

Agreed, general Exceptions are no good -- only did them to get something general in place to start. I think the "class (hierarchy) of PsiturkAPIExceptions to be caught" is the way to go.

> 3. Also, eventually adding logging for command execution might be a good idea.

Agreed.

**Comment by fredcallaway** [comment]
Beyond the print statements and `raw_input` calls (which definitely should be removed), I think the methods of `MTurkServicesWrapper` are still structured in a way that is more natural for the shell than for an API. In particular, I think the methods should be smaller and have fewer arguments. For example, rather than one `worker_approve` method that takes both a hit id and assignment ids, I would have one low-level method that accepts a single worker (a very light wrapper around boto) and a high level method that approves all workers for a single HIT. 

It looks like there is fairly complex logic for handling both arguments at once, but I can't imagine why you would ever want to do this. Also, I'm very surprised by [this comment ](https://github.com/NYUCCL/psiTurk/blob/shell-mturk-abstraction/psiturk/amt_services_wrapper.py#L172). This seems like a post-hoc fix to an upstream bug, in mTurk or in the psiturk database?

**Comment by fredcallaway** [comment]
(As an aside @deargle: how do you get the nice inline code links?)

**Comment by deargle** [comment]
> Beyond the print statements and raw_input calls (which definitely should be removed),

Ah yes, there's one stray `raw_input` in worker_bonus that shouldn't be there. All of the other are in the aws db methods, which I have zero motivation to work on.

> I think the methods of MTurkServicesWrapper are still structured in a way that is more natural for the shell than for an API.

That's a valid criticism. I would still rather we work off of this branch though as we make the wrapper more api-like. The abstraction of all amt_service calls is an important first step which the branch already has done. Plus all the cli speed optimization, which arguably I should have just put straight into master.

> For example, rather than one worker_approve method ... It looks like there is fairly complex logic for handling both arguments at once, but I can't imagine why you would ever want to do this.

Ah, the beauty of inherited code. I'd rather refactor than do a complete rewrite though, so I just ported over what was already there. Open to changing.

> Also, I'm very surprised by this comment .

See this commit message: https://github.com/NYUCCL/psiTurk/commit/9b2b2a43e23635850f3bedb3ebe12586840e5e25

> inline code links?

You just paste in a link to the code range (select the line range by clicking on the line numbers. Press `y` to get a permalink to the branch state). Like this (I surrounded the link with backticks so that it wouldn't render): `https://github.com/NYUCCL/psiTurk/blob/7ae468c809ff7f0982e937e046b657930d5c560c/psiturk/psiturk_shell.py#L222-L283`

**Comment by gureckis** [comment]
re: rds: I believe it could go.  i wrote a lot of that under the idea that people with AMT account automatically have a low cost MySQL service due to AWS but I agree it is likely unused and complicates the code.  this is also not well documented to my memory so its just taking up space.

**Comment by jacob-lee** [comment]
@fredcallaway 
>For example, rather than one worker_approve method ... It looks like there is fairly complex logic for handling both arguments at once, but I can't imagine why you would ever want to do this.

The logic can be simplified (for that, see my example code edit to worker_reject in this thread--ignore the fact that one is unlikely to ever want to reject all workers in a HIT). However, you are right. Instead of, e.g. 

    worker_approve(hitids=[], assignmentids=[])
we have

    approve_assignments(assignmentids=[])

and

    approve_assignments_for_hits(hitids=[])


     



You might take a look at how I changed the code for worker reject (originally posted in thread above):

```python
    def worker_reject(self, hit_ids=[], assignment_ids=[]):
        ''' Reject worker '''
        response = {}
        for hit_id in hit_ids:
            workers = self.amt_services.get_workers('Submitted')
            assignment_ids = assignment_ids + [worker['assignmentId'] for worker in workers if \
                              worker['hitId'] == hit_id]
        for assignment_id in assignment_ids:
            response[assignment_id] = self.amt_services.reject_worker(assignment_id)
        return response
```

**Comment by fredcallaway** [comment]
Something worth considering is whether we want to update the shell code alongside the api code. Obviously we will eventually want to implement the shell with the psiturk API, but for the sake of development, it might make sense to write the API without worrying immediately about updating the shell code to use that API.

Instead of basing the API's structure around the needs of the shell, it might make sense to base it around the needs of a user of the API. To follow this approach, I would start by writing use-case examples, then develop them into unit tests, and finally implement the methods that I decided would be useful to accomplish the needs of the hypothetical user.

In this case, I think the two approaches might lead to the same place, and they may not be entirely conflicting. Perhaps I should write a few use/test cases; then we can think about whether it makes more sense to modify the existing code to fit those cases, or to start "from scratch". Of course, the code can mostly be reused in either case; but starting "from scratch" would give you more freedom to use method names/signatures that make the most sense from a scripting point of view.

**Comment by fredcallaway** [comment]
I agree that the RDS code is cruft. Setting up a database on Heroku is super easy, and self-managing on a server (e.g. EC2) is only slightly more difficult. If we want to make things extremely user-friendly, we could provide scripts (or even API calls?) to create MYSQL databases and tables.

**Comment by deargle** [comment]
Single tear for @gureckis 's RDS code. 😢 RIP

@fredcallaway Use cases are a good idea, especially for the "web dashboard" user.

Given that my branch is technically working, and is progress for the CLI regardless of where we go with the API from here, I think I'll merge it into master unless there are loud objections for some reason.

**Comment by deargle** [comment]
Also @fredcallaway:

> and self-managing on a server (e.g. EC2) is only slightly more difficult

You overestimate a large swath of the psiturk userbase 😉 Or at least, a large swath of the ones who post on the google group.

**Comment by fredcallaway** [comment]
> You overestimate a large swath of the psiturk userbase 
Sounds like an argument for adding automatic database management to the library!

**Comment by deargle** [comment]
Yeah one-click / one-call heroku-psiturk would be slick. Curious about the last comment about Heroku expense. I've gotten by with upgrading to Hobby dyno for when I post hits, at $7 per month. Do you find Hobby dynos not powerful enough for your studies?

**Comment by fredcallaway** [comment]
> Do you find Hobby dynos not powerful enough for your studies?
Actually, it's been fine for me. There have been complaints from others in my lab, but I think they are running extremely large-scale studies (1000s of participants).

**Comment by gureckis** [comment]
i love the idea of the use cases to help flesh out the api first.  it might be interesting to think about what kind of control structures the API could have.  for instance, @alexanderrich and i had discussed if you wanted to repost hits until you reach a particular N in each condition or N in the entire study.  i had thought about that a lot but i wasn't sure how to implement those in a clear way.  i guess there could just be some api calls that do basic checks against your hits/assignments/database and you can check those repeatedly until some condition is met.  this would be very nice and replicable because you can send people a script to "replicate" your design and get rid of any of the psiturk magic that would have to be explained in a documentation file.

**Comment by deargle** [comment]
@gureckis if you're interested I have a tacky batch script that posts N number of hits in batches spaced out over some time period, https://gist.github.com/deargle/f454a156d3a72adf62869e2ef09101d8 But yeah it would be nice to do that in psiturk proper. Or do you mean in the event that you post HITs but they don't all get taken? Something to do that _would_ be helpful, just ran into that for the first time this week.

**Comment by deargle** [comment]
I just merged my branch with the CLI improvements into master, to help with those doing scripting from bash or whatever (like myself and @jacob-lee). I'm not saying the branch is a good API; but it does maintain shell functionality while decoupling the amt_service calls from the shell class, so there's that.

@fredcallaway how do you want to do the use-cases? Collaboratively like on a google doc or something? Or perhaps a wiki page in this repo?

**Comment by fredcallaway** [comment]
For use-cases, I was imagining writing a script that uses the hypothetical psiturk API library. I've been working on this locally. What's the best way to collaborate/share these? I can push to my own branch of psiturk and make a pull request? I'm not sure it makes sense to include these in the main psiturk repo though.

**Comment by gureckis** [comment]
what about https://gist.github.com for snippets

**Comment by deargle** [comment]
Ay, I was getting excited, thought we were going to kick out UML diagrams.

One idea is that we just mirror the `boto` signature calls for `hits` and `workers`, just wrapping those in order to update psiturk statuses. I think the need to update and reference psiturk statuses are the only reason we're not calling boto directly. Also, `hits` and `workers` are really the only objects we care about. Maybe eventually `qualifications`, but that would be new functionality.

github gists are good, although not collaborative-friendly. If you _did_ push to the main repo, we could add code comments. Maybe to a branch? If not just to master.

I made this for notes, if we want to use this too: https://github.com/NYUCCL/psiTurk/wiki/Psiturk-API-use-cases

**Comment by fredcallaway** [comment]
As I've been working on the hypothetical scripting, I'm realizing that an API isn't *really* what I want. It would make things easier than what I'm doing now, but I think even this kind of scripting is more effort than necessary for the typical use case.

I think what @deargle implemented already provides a decent API for programmatic access. So, I think development effort may be better directed towards implementing a higher level interface that obviates the need to do scripting (or manually running many similar commands).

With that in mind, I don't agree that "hits and workers are really the only objects we care about." These are things that psychologists have to think about when working with mturk. But in an ideal world, we wouldn't need to think about HITs. The more relevant high-level organizing unit is an **Experiment** perhaps with different versions.

I should be able to run a command (or click a button) that results in 100 participants taking my study, with ~25 in each of four conditions. In a few hours, I run another command (or check the web interface). I get a prompt saying that the total bonus was calculated to be $XX.XX, and I enter `y` or click confirm. Now I get to start analyzing the data.

On mturk, this could be done by posting one HIT, or posting 12 (mini-batching). But I shouldn't need to think about that. I shouldn't need to keep track of all the HIT ids. I shouldn't need to remember to compensate each one separately, and I shouldn't have to ssh into a server to type in 

```
worker approve --hit AAXXTDSEEATST
worker bonus --auto --hit AAXXTDSEEATST
worker approve --hit ABXRSD123DST
worker bonus --auto --hit ABXRSD123DST
....
```

**Comment by fredcallaway** [comment]
Given the API @deargle implemented, could you create a web interface as a custom route? Is there any reason you couldn't create hits and compensate workers through those routes?

**Comment by deargle** [comment]
Totally possible to implement a custom route to do those things that uses
the _approach_ I wrote, but most web functions are not web-client ready
yet. They still have print functions, they don't throw exceptions yet. Once
that is done, mission-ready, although as you noted, the function signatures
in the wrapper are a bit awkward because they are clones of functions used
by argparse. But they're still usable.

What I meant by saying that those are the only thing that psiturk cares
about is that those are the only things we necessarily need to write
wrappers around boto for, because they rely on info in psiturk db. Other
things we can theoretically call boto directly with. Maybe best to wrap
everything regardless though.

Jacob lee said he would be willing to look into implementing exception
throwing. Removal of print statements can be done at the same time. Still
up for that Jacob?

And another thing. If we want to be able to post batches of hits spread out
over time, we will need to look into some sort of cron functionality, like
rails scheduled tasks.

**Comment by fredcallaway** [comment]
> What I meant by saying that those are the only thing that psiturk cares
about is that those are the only things we necessarily need to write
wrappers around boto for

I see; I wasn't considering database updating. Generally, I have been imagining a high-level API, whereas you have (correctly I think) started with a low-level wrapper around boto that handles the psiturk book-keeping. I identify three tasks that should be completed before we start building higher level functionality (e.g. a web interface) on the API

1. Replace print statements with logging
2. Raising informative exceptions
3. Check for computational efficiency (e.g. don't iterate over all HITs when you can get the relevant HIT(s) with a more precise boto function.)


-------------

> Maybe best to wrap everything regardless though.
I think this is a good idea. It won't be difficult to implement and it will allow users to use a singe library. We might consider returning Python objects (namedtuples or full-fledged classes) rather than dictionaries.

> If we want to be able to post batches of hits spread out over time, we will need to look into some sort of cron functionality

Would the [default scheduling library](https://docs.python.org/3/library/sched.html) suffice?

**Comment by jacob-lee** [comment]
@deargle @fredcallaway 

Your (1),(2), and (3) I think is the main idea. The interface a requester uses will determine how it wants to handle exceptions, etc. Logging, actually, might be best handled by the clients.

Returning objects from a PsiturkResponse class (or some such) is fine too. That is pretty much what boto3 does, and provides an way of getting a dict.

I think everything in boto3 should be wrapped. Seems to me that there are three use-cases for this wrapping:

1. to make common interactions with boto/boto3 more convenient.
2. to abstract between our client code and boto/boto3 (so that changing between boto/boto3, or changes to boto/boto3 by Amazon doesn't induce whole-sale changes in our code, just the wrapper.)
3. to sync/integrate our database info and information retrieved from aws.

It seems like the non-amazon-rds-db code in amt_services.py is really a utility class to make some common interactions with boto more convenient.

**Comment by fredcallaway** [comment]
I think it makes sense to switch to py3 at the same time as we make these structural changes. I would vote to not attempt to maintain compatibility with py2.

Thus, given that we'll be breaking backward compatibility, I think we should be working on a branch other than master.

**Comment by deargle** [comment]
I disagree. I view the web dashboard feature as independent ot py3. There's
really not many structural changes remaining to be made at this point. The
py3 upgrade will be more painful, because of swapping out boto with boto3,
and the new mturk api framework that comes with that. Exception handling in
the wrapper should generalize easily to py3. Web interface upgrade can
happen before and independent of py3. It's just that py3 isn't technically
required for anything we want to do. As for dependent packages dropping py3
support, big deal, we haven't upgraded most packages in eons anyway besides
gunicorn once or twice. There's no sunset for the old mturk api we're on
that boto uses.

Now for the py3 upgrade we may want to follow the pattern of other
projects, including boto, and create a new repo called "psiturk3". Work
there can begin with swapping out boto with boto3, while maintaining all
other functionality. Web dashboard upgrade (which will consist solely of
custom routes and html/js) should be able to port over quite easily.

Who is planning on working on py3 upgrade anyhow?

@gureckis, want to make a psiturk3 repo under NYUCCL?

**Comment by deargle** [comment]
A step towards auto-approving workers: https://github.com/NYUCCL/psiTurk/commit/374d907f3f1b86320802ba083e1691635d682738

See the commit notes though. Currently works by fetching _all workers_ for the current mturk credentials (well, technically, it first fetches all _hit ids_ for the current credentials, and then fetches all workers for all hit_ids). Then the code attempts to approve each assignment_id for the current mturk credentials -- but the inner function does not approve an assignment if the assignment_id is not found in the current psiturk database (e.g., if it's associated with a different study). Could be improved by not attempting to fetch workers for hits not found in the current database. The code currently fetches _all_ hits for the mturk credentials if not given a single hit to fetch workers for.

I think I'll change it so that the `amt_service.get_workers` function can accept a list of `chosen_hits`, and then the `wrapper` will query the psiturk db to get a list of hit_ids if the `--all` argument is passed.

**Begs the question**: should `hit list` continue to list _all hits_ for the current mturk credentials, whether or not a call to `worker approve --hit <hit id>` would work for all of the listed hits? Or should it only fetch hits associated with the current study?

**Comment by deargle** [comment]
I can make it so that the default for `hit list` only returns hits for the current study, while `hit list --all` returns _everything_. I think the "count" for the shell makes more sense to only count hits associated with the current study.

One potential downside of this is if the database happens to get lost. Would be difficult to approve all workers for all hits... wait, hold that, I can add a `--force` flag option to `worker approve --all`. 

So, in summary, I'm proposing doing the following:

* default `hit list` to only return hits for the current study
    * include a `hit list --all` argument
* default `worker approve --all` to only try to approve workers from the current study
* allow `get_workers` to accept a _list_ of chosen_its
* add an argument for `worker approve --all-workers-in-all-studies` which is bordering on ridiculous but whatever, the wrapper makes it all easy.

**Comment by fredcallaway** [comment]
I think this is an _excellent_ idea. This already removes most of the burden of non-automatic approval. You can just set up a recurring CRON job that runs `worker approve -all` and `worker bonus --auto --all`.

For implementation, I imagine that you are thinking something like this?

1. Select unique hit id's from the sql table
2. Collect assignments for each hitid (`get_workers`)
3. Approve each assignment id.

However, I wonder if this might be easier... (and faster?)

1. Select rows in table where status = 4
2. Approve assignment id associated with each row.

**Comment by jacob-lee** [comment]
> default hit list to only return hits for the current study; include a hit list --all argument

If you mean defaulting the `hit list --active` option, then I'm all for it.


> default worker approve --all to only try to approve workers from the current study

In the previous point, the --all flag meant all hits, but here it seems to mean --active. How about just add an --active flag

**Comment by deargle** [comment]
Not quite. By "hit list --all", I meant "hit list --all-studies". By
default, without that flag, it would only return hits related to the
database set in the config file of the directory from where the command was
invoked. Current (unexpected) behavior is for hit list to return all hits
for the given credentials, regardless of whether those hits are associated
with the current directory/study. Whether we switch the default for "hit
list" to really be "hit list --active" is a separate question I wasn't
considering.

And by "worker approve --all", I meant "approve all workers for the current
study".

I'm considering a "worker approve --all --all-studies" which would approve
all workers for the current credentials regardless of whether or not the
associated db entry can be found/updated because the worker might not be
associated with the current study/directory and therefore unknowable to the
current psiturk invocation. But the implications of such a command might
not be obvious, so perhaps it would require a "are you sure you want to do
this?!" confirmation prompt.

**Comment by jacob-lee** [comment]
OIC. 

the problem with `worker approve --all --all-studies` is that the process does not necessarily have the relevant information about which databases or database tables to update for the other studies. to get around that, we'd have to (radically) restructure psiturk away from a one-task paradigm.

**Comment by deargle** [comment]
> the problem with worker approve --all --all-studies is that the process does not necessarily have the relevant information about which databases or database tables to update for the other studies. to get around that, we'd have to (radically) restructure psiturk away from a one-task paradigm.

That's right. But consider the use case of someone whose database has become corrupted or lost, but they want to approve workers anyway. They understand the implications that their database won't be updated, but they don't have a database anymore anyway. They still have to approve, and they don't want to write a custom boto script to approve everyone. 

Maybe that use-case is so fringe that it doesn't matter. But it could easily be built into psiturk though with a simple flag / switch.

**Comment by jacob-lee** [comment]
Not really fringe. We run non-psiturk compensation HITs (because single-task without resorting to smelly hacks). I've approved those non-psiturk HITs using the psiturk interface, sometimes inadvertently... so differentiating the two sounds like a good idea to me.

**Comment by deargle** [comment]
You don't use https://github.com/NYUCCL/whoops_yall for makeup hits?

This is getting silly, but maybe I'll do this

* ` worker approve --all ` | approve all for the current study
* `worker approve --force --hit <hitid>`  | for if you just want to force-approve just one hit not in your current db, but don't want to do all hits everywhere.
* ` worker approve --all --all-studies --force` | approve all workers for all studies

**Comment by jacob-lee** [comment]
The psiturk server is on running our ongoing study 24/7. I would need a
second server, or figure out how to run multiple instances of psiturk
behind the proxy server without hazards and routing collisions.

we're getting into tar-like territory with those opts. (:

**Comment by fredcallaway** [comment]
Here's a fast-running [script](https://gist.github.com/fredcallaway/410aba2642cb037d06f4dbd6f042a8b3) that approves and bonuses all participants in the database.

I (partially) reimplement functionality of `MTurkServicesWrapper`. A key design difference is that my `approve` and `bonus` functions take a single argument: a `Participant` object. I find that this dramatically simplifies the implementation and usage of the functions. Is there a downside to this approach?

A detail: I don't explicitly handle the "multiple participants with one assignment id" problem (as addressed by 9b2b2a4). However, I _think_ that passing in a `Participant` object directly bypasses this problem. If there are multiple `Participant`s with a given `assignmentid`, you may try and fail to approve the extraneous one, but you will also try and (presumably) succeed to approve the correct one. But it is very possible that I don't fully understand this edge case.

**Comment by deargle** [comment]
No need for `MTurkServicesWrapper` in your script, just import and use
`amt_services` directly.

Yes, this approach is lightning fast because it entirely bypasses the need
to query amt for a list of submitted workers. But it does not allow for
workers to be approved if the db has been lost (fringe case).

As for 9b2b2a4, it addresses two problems.

1.

> If there are multiple Participants with a given assignmentid, you may try
and fail to approve the extraneous one, but you will also try and
(presumably) succeed to approve the correct one.

tldr, assuming results are ordered from earliest records to latest, you
won't fail the first time, but you will fail the second time. And it's the
second one that should have gotten the status=5, not the first one (at
least, I think that's the case).

The problem is that, say there are two `assignmentids` in your database
with status == 4, let's say workerid=1 and workerid=2, but in reality only
workerid=2 successfully submitted. (Curse you, mturk submit page, which
doesn't let us know whether a hit was actually successfully submitted so we
have to assume "4" if we ask mturk nicely to do it once), So you query the
psiturk db and get those two workerids with the same assignmentid. You
start with the first, which again, didn't actually successfully submit. You
ask mturk to approve the given assignmentid (curse you, mturk, for only
allowing us to give you an assignmentid to approve, and not also a
workerid). Mturk comes back and says "success I approved that assignment
id." So you say "great, I'll update the status for this worker to 5 *even
though they weren't the actual worker who successfully submitted, that
status actually should have gone to workerid=2*. So then you try to approve
workerid=2, it errors because you already approved him. Database is now
wrong. Doing all that fancy footwork where you query mturk for the workerid
associated with the assignmentid prevents that problem, but the tradeoff is
extra mturk calls. Not a big problem for the original psiturk devs, who I
think did not run studies with super-large n, @gureckis correct me if I'm
wrong (the original code actually never asked for more than 100 workers,
regardless of how many there actually were, but that went unnoticed for a
long time because of small n). I suppose that an alternative to handling
that case is to assume that if there are more than one entries in the
psiturkdb for a given assignmentid with status=4, then assume that the
*later* one is the one to update the psiturk db status for. That might be
fine.

2.

The other think that the current worker_approve implementation handles is "and
sometimes there are 0 (if someone submitted without completeing
the task, for example)." Cheaters. But tbh I've never run into that, but
@alexanderrich you have?

**Comment by fredcallaway** [comment]
> No need for `MTurkServicesWrapper` in your script, just import and use
`amt_services` directly.

The wrapper has some code to pass config values to `MTurkServices` which saved me a bit of effort.

> Database is now wrong.
Ah, I wasn't thinking about this. But I do see why it's desirable to keep the DB accurate in this respect. Here's an idea for a simpler and less resource intensive check: Before approving a participant, compare the workerid to the result of [get_assignment](http://boto.cloudhackers.com/en/latest/ref/mturk.html#boto.mturk.connection.MTurkConnection.get_assignment). If the workerid doesn't match, change the status of that participant to 3.

> The other think that the current worker_approve implementation handles is "and
sometimes there are 0

In my implementation, this person will be left with an unapproved HIT (which will eventually be auto-approved). I think that's probably fine considering it's a rare case.

**Comment by fredcallaway** [comment]
As far as truly _automatic_ approval goes, I got stuck when I realized that this would have to happen after the experiment pop up closes and the worker submits. I imagine there's a way for the server to schedule a backend process to be run in the future. Is anyone familiar with this kind of thing?

**Comment by deargle** [comment]
> Here's an idea for a simpler and less resource intensive check: Before
approving a participant, compare the workerid to the result of
get_assignment
<http://boto.cloudhackers.com/en/latest/ref/mturk.html#boto.mturk.connection.MTurkConnection.get_assignment>.
If the workerid doesn't match, change the status of that participant to 3.

That *is* a good idea. But see the final outstanding concern below.

> In my implementation, this person will be left with an unapproved HIT
(which will eventually be auto-approved). I think that's probably fine
considering it's a rare case.

The problem is that mturk auto-approves submissions after a certain period
of time, I think 30 days. So if the psiturk worker approve code never
queried mturk for a list to compare against, an unaccounted for submission
would never be discovered. I think though that mturkers care a lot about
their reputation so they aren't likely to try to be sneaky... especially if
the options are set to require min 100 approvals and 95% approval rate.

> As far as truly *automatic* approval goes, I got stuck when I realized
that this would have to happen after the experiment pop up closes and the
worker submits. I imagine there's a way for the server to schedule a
backend process to be run in the future. Is anyone familiar with this kind
of thing?

How about Redis? I only have enough exposure to it to know its name.

**Comment by deargle** [comment]
I implemented auto-approve (but not auto-bonus) in the dashboard feature on master
